pkg/autoscaling/calculation.go

```diff
@@ -52,89 +49,399 @@ var (
+	storageUsageTarget := (storageMaxThreshold + storageMinThreshold) / 2
+
+	log.Debug("get storage usage information completed",
+		zap.Float64("totalStorageUsedSize", totalStorageUsedSize),
+		zap.Float64("totalStorageUsedSize", totalStorageCapacity),
+		zap.Float64("totalStorageCapacity", totalStorageCapacity),
+		zap.Float64("storageUsage", storageUsage),
+		zap.Float64("storageMaxThreshold", storageMaxThreshold),
+		zap.Float64("storageMinThreshold", storageMinThreshold))
+
+	if storageUsage > storageMaxThreshold {
+		// generate homogeneous tikv plan
+		resources := getStorageResourcesByComponent(strategy, TiKV)
+		homogeneousTiKVCount := getCountByResourceType(resources, homogeneousTiKVResourceType)
+
+		if resourceMap[homogeneousTiKVResourceType] == strategy.NodeCount || (homogeneousTiKVCount != nil && resourceMap[homogeneousTiKVResourceType] > *homogeneousTiKVCount) {
+			// homogeneous instance number reaches k8s node number or the resource limit,
+			// can not scale out homogeneous instance anymore
+			log.Warn("can not scale out homogeneous instance",
+				zap.Uint64("homogeneous instance number", resourceMap[homogeneousTiKVResourceType]),
+				zap.Uint64("k8s node number", strategy.NodeCount),
+				zap.Uint64p("resource limit", homogeneousTiKVCount),
+			)
+			return nil, nil
 		}
+
+		storageScaleSize := totalStorageUsedSize/storageUsageTarget - totalStorageCapacity
+		storeStorageSize := getStorageByResourceType(resources, homogeneousTiKVResourceType)
+		scaleOutCount := uint64(storageScaleSize)/storeStorageSize + 1
+
+		return getHomogeneousScaleOutPlans(scaleOutCount, uint64(len(instances)), strategy.NodeCount, resources, resourceMap, TiKV), nil
 	}
 
-	return plans
+	return nil, nil
 }
 
-func getPlans(rc *cluster.RaftCluster, querier Querier, strategy *Strategy, component ComponentType) []*Plan {
-	var instances []instance
-	if component == TiKV {
-		instances = filterTiKVInstances(rc)
-	} else {
-		instances = getTiDBInstances(rc.GetEtcdClient())
+func getPrometheusAddress(rc *cluster.RaftCluster) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	resp, err := rc.GetEtcdClient().Get(ctx, prometheusAddressKey)
+	if err != nil {
+		return "", err
+	}
+	if len(resp.Kvs) == 0 {
+		return "", errors.New(fmt.Sprintf("length of the response values of the key %s is 0", prometheusAddressKey))
 	}
 
-	if len(instances) == 0 {
-		return nil
+	address := &Address{}
+	err = json.Unmarshal(resp.Kvs[0].Value, address)
+	if err != nil {
+		return "", err
+	}
+
+	return address.String(), nil
+}
+
+func getTiDBPlans(querier Querier, instances []instance, strategy *Strategy, resourceMap map[string]uint64) ([]*Plan, error) {
+	return getCPUPlans(querier, instances, strategy, resourceMap, TiDB)
+}
+
+func getCPUPlans(querier Querier, instances []instance, strategy *Strategy, resourceMap map[string]uint64, component ComponentType) ([]*Plan, error) {
```

Comment:
How about directly using getTiDBPlans?

Reply:
did you mean change getCPUPlans() function name to getTiDBPlans ?
if yes, I'm afraid we cann't.because getCPUPlans() function not only used by getTiDBPlans() function but also used by getTiKVPlans() function, and calling getTiDBPlans() function in getTiKVPlans() function is meaningless, it doesn't make sense.

Reply:
Oh, you are right.